cpp/src/lance/arrow/utils.cc

```diff
@@ -0,0 +1,102 @@
+//  Copyright 2022 Lance Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+#include "lance/arrow/utils.h"
+
+#include <arrow/result.h>
+#include <fmt/format.h>
+
+#include <string>
+#include <vector>
+
+#include "lance/arrow/type.h"
+
+namespace lance::arrow {
+
+::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> Merge(
+    const std::shared_ptr<::arrow::RecordBatch>& lhs,
+    const std::shared_ptr<::arrow::RecordBatch>& rhs,
+    ::arrow::MemoryPool* pool) {
+  ARROW_ASSIGN_OR_RAISE(auto left_struct, lhs->ToStructArray());
+  ARROW_ASSIGN_OR_RAISE(auto right_struct, rhs->ToStructArray());
+  ARROW_ASSIGN_OR_RAISE(auto struct_arr, Merge(left_struct, right_struct, pool));
+  return ::arrow::RecordBatch::FromStructArray(struct_arr);
+}
+
+::arrow::Result<std::shared_ptr<::arrow::Array>> MergeListArray(
+    const std::shared_ptr<::arrow::Array>& lhs,
+    const std::shared_ptr<::arrow::Array>& rhs,
+    ::arrow::MemoryPool* pool) {
+  assert(is_list(lhs->type()) && is_list(rhs->type()));
+  auto left_list_type = std::static_pointer_cast<::arrow::ListType>(lhs->type());
+  auto right_list_type = std::static_pointer_cast<::arrow::ListType>(rhs->type());
+  if (!is_struct(left_list_type->value_type()) || !is_struct(right_list_type->value_type())) {
+    return ::arrow::Status::Invalid(fmt::format(
+        "Can only merge list of structs: left={} right={}", left_list_type, right_list_type));
+  }
+  auto left_list = std::static_pointer_cast<::arrow::ListArray>(lhs);
+  auto right_list = std::static_pointer_cast<::arrow::ListArray>(rhs);
+  auto left_values = std::static_pointer_cast<::arrow::StructArray>(left_list->values());
+  auto right_values = std::static_pointer_cast<::arrow::StructArray>(right_list->values());
+  ARROW_ASSIGN_OR_RAISE(auto values, Merge(left_values, right_values, pool));
+  return ::arrow::ListArray::FromArrays(*left_list->offsets(), *values, pool);
```

Comment:
is it safe to assume the offsets are the same?

Reply:
For the currently limited use case, this supposes to be the case.  

Was hesitant to add an offset comparision for every read tho.